pkg/upstream/cluster/loadbalancer.go

```diff
@@ -624,3 +626,171 @@ func (lb *reqRoundRobinLoadBalancer) IsExistsHosts(metadata api.MetadataMatchCri
 func (lb *reqRoundRobinLoadBalancer) HostNum(metadata api.MetadataMatchCriteria) int {
 	return lb.hosts.Size()
 }
+
+type peakEwmaLoadBalancer struct {
+	info  types.ClusterInfo
+	hosts types.HostSet
+	rand  *rand.Rand
+	mutex sync.Mutex
+
+	wrrLoadBalancer types.LoadBalancer
+	fallback        bool
+
+	choice uint32
+}
+
+func newPeakEwmaLoadBalancer(info types.ClusterInfo, hosts types.HostSet) types.LoadBalancer {
+	fallback := false
+	if hosts.Size() > 1 {
+		hosts.Range(func(host types.Host) bool {
+			// Check whether `UpstreamRequestDuration` is enabled,
+			// or fallback to WRR because all hosts have same duration.
+			if _, ok := host.HostStats().UpstreamRequestDuration.(gometrics.NilHistogram); ok {
+				fallback = true
+				return false
+			}
+			return true
+		})
+	}
+
+	lb := &peakEwmaLoadBalancer{
+		info:            info,
+		hosts:           hosts,
+		rand:            rand.New(rand.NewSource(time.Now().UnixNano())),
+		choice:          default_choice,
+		wrrLoadBalancer: newWRRLoadBalancer(info, hosts),
+		fallback:        fallback,
+	}
+	return lb
+}
+
+func (lb *peakEwmaLoadBalancer) ChooseHost(context types.LoadBalancerContext) types.Host {
+	if lb.fallback {
+		return lb.wrrLoadBalancer.ChooseHost(context)
+	}
+
+	hs := lb.hosts
+	total := hs.Size()
+
+	if total == 0 {
+		return nil
+	}
+	if total == 1 {
+		host := hs.Get(0)
+		if host.Health() {
+			return host
+		}
+		return nil
+	}
+
+	var candidate types.Host
+
+	// If `total` is less than or equal to `choice`, we can iterate over all elements directly.
+	if total <= int(lb.choice) {
+		candidate = lb.iterateChoose()
+	} else {
+		candidate = lb.randomChoose()
```

Comment:
The random choose assumes every host has the same weight. Maybe we can add a check and use it only for unweight branch?

Reply:
I may have misunderstood you.

In fact, neither random nor iterative, they do not distribute traffic according to weight. Iterative selection is just an optimization based on the fact that random selection is based on the fact that if n+k elements (k>0) are selected among n elements, then at least one element will be repeatedly selected (pigeon's nest principle), and for uniqueness and To select elements without repetition, it is to traverse all elements. So iterative selection is just a random selection of an optimization branch

Reply:
So the weight here is not the same as the weight in the WRR? There will likely be an issue: the traffic distributions are different between random and iterative when the hosts have different weights. Each host is selected equally in random choose, so a large weight may not take effect as the probability of being selected is different from the weight. Maybe you can write a script with very imbalanced weights (like 1:10) to see if it is right or not.